src/header.rs

```diff
@@ -0,0 +1,111 @@
+//! headers are sets of consecutive keywords and tokens, such as
+//! `pub const unsafe fn foo` and `pub(crate) unsafe trait Bar`.
+//!
+//! This module contains general logic for formatting such headers,
+//! where they are always placed on a single line except when there
+//! are comments between parts of the header.
+
+use std::borrow::Cow;
+
+use rustc_ast as ast;
+use rustc_span::Span;
+use rustc_span::symbol::Ident;
+use tracing::debug;
+
+use crate::comment::{combine_strs_with_missing_comments, contains_comment};
+use crate::rewrite::RewriteContext;
+use crate::shape::Shape;
+use crate::utils::rewrite_ident;
+
+pub(crate) fn format_header(
+    context: &RewriteContext<'_>,
+    shape: Shape,
+    parts: Vec<HeaderPart>,
+) -> String {
+    debug!(?parts, "format_header");
+    let shape = shape.infinite_width();
+
+    // Empty `HeaderPart`s are ignored.
+    let mut parts = parts.into_iter().filter(|x| !x.snippet.is_empty());
+    let Some(part) = parts.next() else {
+        return String::new();
+    };
+
+    let mut result = part.snippet.into_owned();
+    let mut span = part.span;
+
+    for part in parts {
+        debug!(?result, "before combine");
+        let comments_span = span.between(part.span);
+        let comments_snippet = context.snippet(comments_span);
+        result = if contains_comment(comments_snippet) {
+            // FIXME(fee1-dead): preserve (potentially misaligned) comments instead of reformatting
+            // them. Revisit this once we have a strategy for properly dealing with them.
+            format!("{result}{comments_snippet}{}", part.snippet)
+        } else {
+            combine_strs_with_missing_comments(
+                context,
+                &result,
+                &part.snippet,
+                comments_span,
+                shape,
+                true,
+            )
+            .unwrap_or_else(|_| format!("{} {}", &result, part.snippet))
+        };
+        debug!(?result);
+        span = part.span;
+    }
+
+    result
+}
+
+#[derive(Debug)]
+pub(crate) struct HeaderPart {
+    /// snippet of this part without surrounding space
+    snippet: Cow<'static, str>,
+    span: Span,
+}
+
+impl HeaderPart {
+    pub(crate) fn new(snippet: impl Into<Cow<'static, str>>, span: Span) -> Self {
+        Self {
+            snippet: snippet.into(),
+            span,
+        }
+    }
+
+    pub(crate) fn ident(context: &RewriteContext<'_>, ident: Ident) -> Self {
+        Self {
+            snippet: rewrite_ident(context, ident).to_owned().into(),
+            span: ident.span,
+        }
```

Comment:
Should this be implemented in terms of `HeaderPart::new`? Would save you needing to call `.into()`.

```suggestion
        let snippet = rewrite_ident(context, ident).to_owned();
        Self::new(snippet, ident.span)
```

If we can make `HeaderPart` hold a `Cow<'a, str>`, then this could probably be written as:

```rust
Self::new(rewrite_ident(context, ident), ident.span)
```